### PR TITLE
MySQL Alpha Support

### PR DESCRIPTION
**Pull request by Rentacookie**
### Alpha support for MySQL as a replication datasource

Replication for MySQL is handled by listening for changes on the [binlog](https://dev.mysql.com/doc/refman/8.4/en/binary-log.html) which is similar to Postgres' write ahead log.

We use a forked version of the Zongji MySQL Binlog listener library that has been customised to our needs:
https://github.com/powersync-ja/powersync-mysql-zongji

Demo app and backend is available here:
https://github.com/powersync-ja/self-host-demo/blob/main/demos/nodejs-mysql/README.md

Some additional configuration is necessary for your instance of MySQL to enable the binlog. A MySQL user should also be created with the required permissions for Powersync to perform the replication.
https://github.com/powersync-ja/self-host-demo/tree/main/services/mysql/init-scripts

### CODE REVIEW

File: modules/module-mysql/src/replication/zongji/zongji.d.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
+declare module '@powersync/mysql-zongji' {
```
Comment (stevensJourney) [inline modules/module-mysql/src/replication/zongji/zongji.d.ts:1]: Should we move this to our fork package?

File: packages/service-core/src/util/protocol-types.ts [hunk 2]
```diff
@@ -88,7 +88,7 @@ export type StreamingSyncLine =
  */
 export type OpId = string;
 
-export interface Checkpoint {
+interface Checkpoint {
```
Comment (stevensJourney) [inline packages/service-core/src/util/protocol-types.ts:91]: Why this change?

File: modules/module-mysql/src/replication/zongji/zongji.d.ts (hunk 1 above)
Comment (Rentacookie) [reply modules/module-mysql/src/replication/zongji/zongji.d.ts:1]: I would be happy to - just not entirely sure what configuration is involved for that 😅 
I'm assuming this entails exporting it as the library's types or something?

Comment (stevensJourney) [reply modules/module-mysql/src/replication/zongji/zongji.d.ts:1]: This file could be added to that package. It would require some minor modifications, like removing the `declare module ...` bit and exporting the individual declarations. We could then add an entry for `"types": "types/index.d.ts` to the package.json file. 
This is not a blocker for now though.

**Review by stevensJourney — APPROVED** [review]
Looks good from my side. We might need to cleanup some of the TODO comments and track those items externally.

**Review by stevensJourney — APPROVED** [review]
Looks good to me

File: modules/module-mysql/dev/config/sync_rules.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# See Documentation for more information:
+# https://docs.powersync.com/usage/sync-rules
+# Note that changes to this file are not watched.
+# The service needs to be restarted for changes to take effect.
+
+# Note that specifying the schema is currently required due to the default
+# schema being specified as `public`, but in mysql the schema is the database name
```
Comment (rkistner) [inline modules/module-mysql/dev/config/sync_rules.yaml:7]: Is this still the case, with `defaultSchema` now being the database name?

File: modules/module-mysql/test/src/util.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  const version = await getMySQLVersion(connection);
+  if (gte(version, '8.4.0')) {
+    await connection.query('RESET BINARY LOGS AND GTIDS');
+  } else {
+    await connection.query('RESET MASTER');
+  }
+
+  // await connection.query(`DROP DATABASE IF EXISTS ${TEST_CONNECTION_OPTIONS.database}`);
+  //
+  // await connection.query(`CREATE DATABASE IF NOT EXISTS ${TEST_CONNECTION_OPTIONS.database}`);
```
Comment (rkistner) [inline modules/module-mysql/test/src/util.ts:50]: Can these comments be removed?

**Review by rkistner — COMMENTED** [review]
Happy on the high-level structure, will see if I can still go through the details of the replication process at some point

File: modules/module-mysql/dev/config/sync_rules.yaml (hunk 3 above)
Comment (Rentacookie) [reply modules/module-mysql/dev/config/sync_rules.yaml:7]: No you are right, no longer needed :)

File: modules/module-mysql/package.json [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      "require": "./dist/types/types.js",
+      "default": "./dist/types/types.js"
+    }
+  },
+  "dependencies": {
+    "@powersync/lib-services-framework": "workspace:*",
+    "@powersync/service-core": "workspace:*",
+    "@powersync/service-sync-rules": "workspace:*",
+    "@powersync/service-types": "workspace:*",
+    "@powersync/mysql-zongji": "0.0.0-dev-20241023144335",
```
[the remark this answers is not in the document]
Comment (Rentacookie) [reply modules/module-mysql/package.json:35]: Definitely. I'm still struggling with an issue where the testing framework of that zongji lib causes the build to fail even though all the tests pass. That is currently the main blocker for getting out a proper released version.
